### Question 1

Explain succinctly in your own words (but working with a ChatBot if needed)...

- Q: the difference between Simple Linear Regression and Multiple Linear Regression; and the benefit the latter provides over the former. 

A: As opposed to Simple Linear Regression, Multiple Linear Regression includes more than one independent variables. This means that multiple linear regression models can capture the influence of multiple factors on the outcome, which is often more realistic and applicable to complex scenarios where multiple factors are at play.
    
    
- Q: the difference between using a continuous variable and an indicator variable in Simple Linear Regression; and these two linear forms

A: With a continuous variable, the relationship is a straight line with a defined slope and intercept. With an indicator variable, the model consists of two points or levels representing the mean outcome for each group, connected by the difference beta_1.

- the change that happens in the behavior of the model (i.e., the expected nature of the data it models) when a single indicator variable is introduced alongside a continuous variable to create a Multiple Linear Regression; and these two linear forms (i.e., the Simple Linear Regression versus the Multiple Linear Regression)

A: Without the indicator variable, the model would fit a single regression line for the relationship between x and y, assuming this relationship holds across all observations. By adding the indicator variable, the model can now represent two distinct lines for the two groups: 1. These lines are parallel (since they share the same slope beta_2), and 2. They have different intercepts, separated by beta_1.

Linear forms: 

y=β 
0
​
 +β 
1
​
 D+β 
2
​
 x+ϵ
where:


β0: is the intercept for the baseline group (when 𝐷=0),

β1: represents the shift in the intercept for the other group (when D=1),

β2: is the slope of the continuous variable x, shared across both groups.

- the effect of adding an interaction between a continuous and an indicator variable in Multiple Linear Regression models; and this linear form

Without the interaction, the model would produce two parallel lines, one for each group, differing only in intercept but not in slope. With the interaction term, however: Each group has its own intercept and slope. The lines are no longer parallel, meaning the relationship between the continuous predictor x and the outcome y varies depending on the group defined by D.

Linear forms:

y=β 
0
​
 +β 
1
​
 D+β 
2
​
 x+β 
3
​
 (D×x)+ϵ
where:

𝛽0: The intercept for the baseline group (when D=0),

𝛽1: The difference in intercept between the two groups (the shift in intercept for 𝐷=1),

𝛽2: The slope of x for the baseline group (when 𝐷=0),

𝛽3: The interaction term, capturing the difference in the slope of x between the two groups.

- the behavior of a Multiple Linear Regression model (i.e., the expected nature of the data it models) based only on indicator variables derived from a non-binary categorical variable; this linear form; and the necessarily resulting binary variable encodings it utilizes

the model essentially partitions the data into different groups, each with its own mean or intercept, while not modeling any continuous or linear trend. The expected nature of the data in such a model is therefore group-based, with a separate mean response for each category.

linear form:

y=β0 + β1​Indicator_1 + β2​Indicator_2 + βn​Indicator_3 + ϵ

### Question 2

Outcome and Predictor Variables
Outcome Variable:

The primary outcome we want to predict is Sales (denoted 
𝑦
y), which represents the revenue or number of items sold as a result of the advertising efforts.
Predictor Variables:

TV Advertising Spend (denoted 
TV_Spend
TV_Spend): The amount of money the company spends on TV ads.
Online Advertising Spend (denoted 
Online_Spend
Online_Spend): The amount of money the company spends on online ads.
Potential for Meaningful Interactions
An interaction effect would be meaningful here if the effect of TV advertising spend on sales changes depending on how much is spent on online advertising, and vice versa. This could happen if the two forms of advertising complement each other. For example:

When both advertising channels are used together, the impact on sales might be greater than using each one in isolation.
Conversely, spending heavily on one type of ad might reduce the need for or effectiveness of spending on the other.
In such a case, including an interaction term between TV and online advertising spend would allow the model to capture this combined effect, giving a more accurate prediction of sales.

Linear Forms of the Model
1. Without Interaction:
Without an interaction term, we assume that TV and online advertising affect sales independently of each other. The linear form is: 𝑦 = 𝛽0 + 𝛽1⋅TV_Spend + 𝛽2⋅Online_Spend + 𝜖

where:

𝛽
0
​
 : The baseline sales when there is no spending on either advertising channel.
𝛽
1
​
 : The effect of each additional dollar spent on TV advertising on sales, assuming the online advertising spend remains constant.
𝛽
2
​
 : The effect of each additional dollar spent on online advertising on sales, assuming the TV advertising spend remains constant.
In this model, TV and online advertising have separate, additive effects on sales. The combined impact of TV and online ads is simply the sum of the two individual effects, and we do not account for any synergy or diminishing returns between the two.

2. With Interaction:
Including an interaction term allows the effect of TV advertising on sales to depend on the amount spent on online advertising, and vice versa. The linear form with an interaction is: 𝑦 = 𝛽0 + 𝛽1⋅TV_Spend + 𝛽2⋅Online_Spend + 𝛽3⋅(V_Spend×Online_Spend) + 𝜖

where:

𝛽
0
​
 : The baseline sales when there is no spending on either advertising channel.
𝛽
1
​
 : The effect of each additional dollar spent on TV advertising when the online advertising spend is zero.
𝛽
2
​
 : The effect of each additional dollar spent on online advertising when the TV advertising spend is zero.
𝛽
3
​
 : The interaction effect, representing how the effectiveness of TV spending on sales changes with each dollar spent on online advertising, and vice versa.
 
 
 Sub-Questions:
 
1. 
- Without interaction: Substitute the values of TV_Spend and Online_Spend with the amount spent on each type of advertising. Use the estimated coefficients 𝛽0, 𝛽1, and 𝛽2 obtained from fitting the model to the data. Calculate 𝑦 (predicted sales) by adding the contributions from each term.

- With interaction: Substitute the values of TV_Spend and Online_Spend with the amounts spent on each type of advertising. Use the estimated coefficients 𝛽0, 𝛽1, 𝛽2, and 𝛽3. Calculate the interaction term by multiplying TV_Spend and Online_Spend, then multiply by 𝛽3. Add up all terms to get the predicted 𝑦 (sales).

2. 
- Without Interaction: The predictions assume that the effects of TV and online ad spends on sales are independent. The two advertising channels (TV and online) are treated as additive effects, meaning that their combined effect on sales is simply the sum of their individual effects. There’s no consideration of how the two types of advertising might interact with each other.

- With Interaction: The predictions now account for the possibility that the effect of one advertising medium on sales could change depending on the level of spending on the other. The interaction term allows for the combined effect of high TV and high online spending to be different from the sum of their individual effects, capturing synergy (or even diminishing returns) when both channels are used together.

Chat summary and link for Q 1 and 2:

https://chatgpt.com/share/67366311-2888-800c-9ecb-adc3c63a53b0

1. Simple vs. Multiple Linear Regression:
Simple Linear Regression involves predicting an outcome based on one predictor variable, assuming a linear relationship between them.
Multiple Linear Regression uses two or more predictor variables to predict the outcome, allowing for the modeling of more complex relationships.
2. Continuous vs. Indicator Variables in Regression:
In Simple Linear Regression, a continuous predictor variable can be used directly to predict the outcome, while an indicator (dummy) variable represents categorical data with 0s and 1s to capture the presence or absence of a characteristic.
Multiple Linear Regression allows for both continuous and indicator variables as predictors. Adding an interaction term between a continuous and indicator variable allows us to model how the effect of one predictor changes depending on the level of another.
3. Interaction Effects:
Interaction between predictors occurs when the effect of one predictor on the outcome depends on the value of another predictor. For example, an interaction between TV and online advertising spend might capture how the impact of TV ads on sales changes depending on online spending (and vice versa).
Including interaction terms in regression models helps capture combined effects that might not be observed when predictors are considered independently.
4. Categorized (Binary) Variables in Regression:
When predictor variables are binary (e.g., "high" or "low" spending on TV or online ads), we use indicator variables. The regression model then estimates the effect of "high" spending relative to "low" spending, with or without interaction.
The model without interaction assumes that the effect of each advertising type is independent. The model with interaction allows the effects to depend on each other, capturing a more complex relationship (e.g., synergistic effects of combined spending).
5. Predictions from Models:
In models without interaction, predictions are based on the additive effects of predictors, assuming no synergy between them.
In models with interaction, predictions reflect how one predictor’s effect on the outcome depends on the level of the other predictor, providing a more nuanced understanding of the combined impact of multiple variables.
This covers the core concepts and discussions about regression, including continuous, indicator, and interaction variables, as well as how to interpret and make predictions from such models. Let me know if you’d like further clarification on any point!

In [1]:
# Question 4

import pandas as pd

url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
# fail https://github.com/KeithGalli/pandas/blob/master/pokemon_data.csv
pokeaman = pd.read_csv(url) 
pokeaman

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [2]:
# Question 4

import statsmodels.formula.api as smf

model1_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation) + Q("Sp. Def"):C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") * C(Generation)', data=pokeaman)

model2_fit = model2_spec.fit()
model2_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.176
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     15.27
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           3.50e-27
Time:                        21:44:52   Log-Likelihood:                -3649.4
No. Observations:                 800   AIC:                             7323.
Df Residuals:                     788   BIC:                             7379.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===================================================================================================
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          26.8971      5.246      5.127      0.000      16.599      37.195
C(Generation)[T.2]                 20.0449      7.821      2.563      0.011       4.692      35.398
C(Generation)[T.3]                 21.3662      6.998      3.053      0.002       7.629      35.103
C(Generation)[T.4]                 31.9575      8.235      3.881      0.000      15.793      48.122
C(Generation)[T.5]                  9.4926      7.883      1.204      0.229      -5.982      24.968
C(Generation)[T.6]                 22.2693      8.709      2.557      0.011       5.173      39.366
Q("Sp. Def")                        0.5634      0.071      7.906      0.000       0.423       0.703
Q("Sp. Def"):C(Generation)[T.2]    -0.2350      0.101     -2.316      0.021      -0.434      -0.036
Q("Sp. Def"):C(Generation)[T.3]    -0.3067      0.093     -3.300      0.001      -0.489      -0.124
Q("Sp. Def"):C(Generation)[T.4]    -0.3790      0.105     -3.600      0.000      -0.586      -0.172
Q("Sp. Def"):C(Generation)[T.5]    -0.0484      0.108     -0.447      0.655      -0.261       0.164
Q("Sp. Def"):C(Generation)[T.6]    -0.3083      0.112     -2.756      0.006      -0.528      -0.089
==============================================================================
Omnibus:                      337.229   Durbin-Watson:                   1.505
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2871.522
Skew:                           1.684   Prob(JB):                         0.00
Kurtosis:                      11.649   Cond. No.                     1.40e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Question 4:

The apparent contradiction arises because R-squared is a measure of overall model fit (how much of the total variability is explained by the predictors), while statistical significance (p-values) tests whether a relationship exists between a specific predictor and the dependent variable, regardless of the overall model's explanatory power. In this case, while the model only explains a small portion of the variance, the individual predictors (like Generation and Sp. Def) have strong, statistically significant relationships with the outcome. However, there are likely many other unmeasured factors or complexities that the current model does not capture, which limits the overall R-squared.

moreover:
- Effect Size vs. Explained Variance: The contradiction arises because statistical significance (p-value) is about the reliability of an effect, whereas R-squared is about the amount of variance explained by the model. You can have large, statistically significant coefficients (indicating a strong relationship) but still have a low R-squared because:

    - Large Coefficients: The effect of a predictor might be large (e.g., a 20-unit difference in HP between two generations), but if there are many other factors that influence HP (which aren’t in the model), the R-squared will still be low. The model might be missing other important predictors that could explain more variability in HP.

    - Residual Variability: Even if some predictors have strong effects on HP, other random or unmeasured factors (or noise) could still explain a large portion of the remaining variability, leading to a low R-squared.

In [3]:
import numpy as np
from sklearn.model_selection import train_test_split

fifty_fifty_split_size = int(pokeaman.shape[0]*0.5)

# Replace "NaN" (in the "Type 2" column with "None")
pokeaman.fillna('None', inplace=True)

np.random.seed(130)
pokeaman_train,pokeaman_test = \
  train_test_split(pokeaman, train_size=fifty_fifty_split_size)
pokeaman_train

#                     Name   Type 1   Type 2  HP  Attack  Defense  \
370  338                  Solrock     Rock  Psychic  70      95       85   
6      6                Charizard     Fire   Flying  78      84       78   
242  224                Octillery    Water     None  75     105       75   
661  600                    Klang    Steel     None  60      80       95   
288  265                  Wurmple      Bug     None  45      45       35   
..   ...                      ...      ...      ...  ..     ...      ...   
522  471                  Glaceon      Ice     None  65      60      110   
243  225                 Delibird      Ice   Flying  45      55       45   
797  720      HoopaHoopa Confined  Psychic    Ghost  80     110       60   
117  109                  Koffing   Poison     None  40      65       95   
409  373  SalamenceMega Salamence   Dragon   Flying  95     145      130   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
370       55       65     70           3      False  
6        109       85    100           1      False  
242      105       75     45           2      False  
661       70       85     50           5      False  
288       20       30     20           3      False  
..       ...      ...    ...         ...        ...  
522      130       95     65           4      False  
243       65       45     75           2      False  
797      150      130     70           6       True  
117       60       45     35           1      False  
409      120       90    120           3      False  

[400 rows x 12 columns]

In [4]:
model_spec3 = smf.ols(formula='HP ~ Attack + Defense', 
                      data=pokeaman_train)
model3_fit = model_spec3.fit()
model3_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        21:44:53   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     42.5882      3.580     11.897      0.000      35.551      49.626
Attack         0.2472      0.041      6.051      0.000       0.167       0.327
Defense        0.1001      0.045      2.201      0.028       0.011       0.190
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         343.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [5]:
yhat_model3 = model3_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model3_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model3)[0,1]**2)

'In sample' R-squared:     0.14771558304519894
'Out of sample' R-squared: 0.21208501873920738


In [12]:
model4_linear_form = 'HP ~ Attack * Defense * Speed * Legendary'
model4_linear_form += ' * Q("Sp. Def") * Q("Sp. Atk")'
# DO NOT try adding '* C(Generation) * C(Q("Type 1")) * C(Q("Type 2"))'
# That's 6*18*19 = 6*18*19 possible interaction combinations...
# ...a huge number that will blow up your computer

model4_spec = smf.ols(formula=model4_linear_form, data=pokeaman_train)
model4_fit = model4_spec.fit()
model4_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.467
Model:                            OLS   Adj. R-squared:                  0.369
Method:                 Least Squares   F-statistic:                     4.764
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           4.23e-21
Time:                        21:46:54   Log-Likelihood:                -1738.6
No. Observations:                 400   AIC:                             3603.
Df Residuals:                     337   BIC:                             3855.
Df Model:                          62                                         
Covariance Type:            nonrobust                                         
====================================================================================================================================
                                                                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------------
Intercept                                                          521.5715    130.273      4.004      0.000     265.322     777.821
Legendary[T.True]                                                   -6.1179      2.846     -2.150      0.032     -11.716      -0.520
Attack                                                              -8.1938      2.329     -3.518      0.000     -12.775      -3.612
Attack:Legendary[T.True]                                         -1224.9610    545.105     -2.247      0.025   -2297.199    -152.723
Defense                                                             -6.1989      2.174     -2.851      0.005     -10.475      -1.923
Defense:Legendary[T.True]                                         -102.4030     96.565     -1.060      0.290    -292.350      87.544
Attack:Defense                                                       0.0985      0.033      2.982      0.003       0.034       0.164
Attack:Defense:Legendary[T.True]                                    14.6361      6.267      2.336      0.020       2.310      26.963
Speed                                                               -7.2261      2.178     -3.318      0.001     -11.511      -2.942
Speed:Legendary[T.True]                                            704.8798    337.855      2.086      0.038      40.309    1369.450
Attack:Speed                                                         0.1264      0.038      3.351      0.001       0.052       0.201
Attack:Speed:Legendary[T.True]                                       5.8648      2.692      2.179      0.030       0.570      11.160
Defense:Speed                                                        0.1026      0.039      2.634      0.009       0.026       0.179
Defense:Speed:Legendary[T.True]                                     -6.9266      3.465     -1.999      0.046     -13.742      -0.111
Attack:Defense:Speed                                                -0.0016      0.001     -2.837      0.005      -0.003      -0.001
Attack:Defense:Speed:Legendary[T.True]                              -0.0743      0.030     -2.477      0.014      -0.133      -0.015
Q("Sp. Def")                                                        -5.3982      1.938     -2.785      0.006      -9.211      -1.586
Legendary[T.True]:Q("Sp. Def")                                    -282.2496    126.835     -2.225      0.027    -531.738     -32.761
Attack:Q("Sp. Def")                                                  0.1094      0.034      3.233      0.001       0.043       0.176
Attack:Legendary[T.True]:Q("Sp. Def")                               12.6503      5.851      2.162      0.031       1.141      24.160
Defense:Q("Sp. Def")    

In [13]:
yhat_model4 = model4_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model4_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model4)[0,1]**2)

'In sample' R-squared:     0.46709442115833855
'Out of sample' R-squared: 0.002485342598992873


In [19]:
# Here's something a little more reasonable...
model6_linear_form = 'HP ~ Attack + Speed + Q("Sp. Def") + Q("Sp. Atk")'
# And here we'll add the significant indicators from the previous model
# https://chatgpt.com/share/81ab88df-4f07-49f9-a44a-de0cfd89c67c
model6_linear_form += ' + I(Q("Type 1")=="Normal")'
model6_linear_form += ' + I(Q("Type 1")=="Water")'
model6_linear_form += ' + I(Generation==2)'
model6_linear_form += ' + I(Generation==5)'

model6_spec = smf.ols(formula=model6_linear_form, data=pokeaman_train)
model6_fit = model6_spec.fit()
model6_fit.summary()

yhat_model6 = model6_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model6_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2)

'In sample' R-squared:     0.3326310334310908
'Out of sample' R-squared: 0.29572460427079933


### Question 5:

what do each of the cells do?
cell 1: this code splits the pokeaman dataset into two equal halves, replacing missing values with 'None' and setting up a reproducible random split.

cell 2: this code fits a linear regression model to predict HP based on Attack and Defense, and provides a summary of the results.

cell 3: The first part prints the R-squared for the training data (in-sample), which tells how well the model fits the training data.The second part prints the R-squared for the test data (out-of-sample), which tells how well the model generalizes to unseen data, based on the correlation between actual and predicted values from the test set.

cell 4: This code builds a linear regression model with HP as the dependent variable and a large number of independent variables and interaction terms, including both main effects and complex interactions. The formula dynamically incorporates multiple variables and interactions, but it avoids a massive computational burden by excluding combinations with certain categorical variables (Generation, Type 1, Type 2). The model is then fitted and a summary of the results is displayed.

cell 5: In-sample R-squared (model4_fit.rsquared) measures how well the model fits the training data.
Out-of-sample R-squared is computed by calculating the correlation between actual and predicted HP values from the test data, then squaring that correlation to obtain a value that shows how well the model generalizes to new, unseen data.

### Question 6:

The design matrix (model4_spec.exog) is constructed based on the formula you specify (model4_linear_form). It transforms the original dataset by creating columns for each predictor, including both the main effects and any interaction terms. This matrix is then used in the linear regression model to predict the outcome variable (HP), by performing a linear combination of the predictors and their coefficients.

Multicollinearity contributes to a lack of out-of-sample generalization by causing instability in the model's coefficients and inflating their variances. This makes the model highly sensitive to small changes in the training data, leading to overfitting and poor generalization to new, unseen data (as indicated by low out-of-sample R-squared). Essentially, the model fails to learn the true underlying relationships and instead "memorizes" the correlated patterns in the training set, which do not apply to the test data.

### Question 7:

- Q: how is model 5 taken from model 3 and 4?

A: It builds on Model 3 by incorporating more relevant predictors (like Pokémon types, generation, and special stats) to improve the model’s explanatory power. It also improves on Model 4 by simplifying the model (avoiding overly complex interactions) and addressing multicollinearity issues, making the model more stable and generalizable.

- Q: how is model 6 taken from model 5?

A: Model 6 extends model5 by adding categorical predictors, particularly those related to Pokémon types and generations, that have been identified as important.

- Q how is model 7 taken from model 6?

A: Model 7 improves upon Model 6 by addressing key issues related to scaling and numerical stability, especially when interaction terms are present. Through centering and scaling, Model 7 reduces multicollinearity, improves model stability, and provides more interpretable coefficients.

In [16]:
# And here's a slight change that seems to perhaps improve prediction...
model7_linear_form = 'HP ~ Attack * Speed * Q("Sp. Def") * Q("Sp. Atk")'
model7_linear_form += ' + I(Q("Type 1")=="Normal")'
model7_linear_form += ' + I(Q("Type 1")=="Water")'
model7_linear_form += ' + I(Generation==2)'
model7_linear_form += ' + I(Generation==5)'

model7_spec = smf.ols(formula=model7_linear_form, data=pokeaman_train)
model7_fit = model7_spec.fit()
model7_fit.summary()

yhat_model7 = model7_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model7_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2)

# And here's a slight change that seems to perhas improve prediction...
model7_linear_form_CS = 'HP ~ scale(center(Attack)) * scale(center(Speed))'
model7_linear_form_CS += ' * scale(center(Q("Sp. Def"))) * scale(center(Q("Sp. Atk")))'
# We DO NOT center and scale indicator variables
model7_linear_form_CS += ' + I(Q("Type 1")=="Normal")'
model7_linear_form_CS += ' + I(Q("Type 1")=="Water")'
model7_linear_form_CS += ' + I(Generation==2)'
model7_linear_form_CS += ' + I(Generation==5)'

model7_CS_spec = smf.ols(formula=model7_linear_form_CS, data=pokeaman_train)
model7_CS_fit = model7_CS_spec.fit()
model7_CS_fit.summary().tables[-1] 
# "Cond. No." is NOW 15.4 due to centering and scaling

'In sample' R-squared:     0.37818209127432456
'Out of sample' R-squared: 0.35055389205977444


<class 'statsmodels.iolib.table.SimpleTable'>

In [17]:
model7_gen1_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model7_gen1_predict_future_fit = model7_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model7_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

'In sample' R-squared:     0.37818209127432456 (original)
'Out of sample' R-squared: 0.35055389205977444 (original)
'In sample' R-squared:     0.5726118179916575 (gen1_predict_future)
'Out of sample' R-squared: 0.11151363354803218 (gen1_predict_future)


In [18]:
model7_gen1to5_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model7_gen1to5_predict_future_fit = model7_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model7_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

'In sample' R-squared:     0.37818209127432456 (original)
'Out of sample' R-squared: 0.35055389205977444 (original)
'In sample' R-squared:     0.3904756578094535 (gen1to5_predict_future)
'Out of sample' R-squared: 0.23394915464343125 (gen1to5_predict_future)


In [20]:
model6_gen1_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model6_gen1_predict_future_fit = model6_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model6_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

'In sample' R-squared:     0.3326310334310908 (original)
'Out of sample' R-squared: 0.29572460427079933 (original)
'In sample' R-squared:     0.4433880517727282 (gen1_predict_future)
'Out of sample' R-squared: 0.1932858534276128 (gen1_predict_future)


In [21]:
model6_gen1to5_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model6_gen1to5_predict_future_fit = model6_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model6_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

'In sample' R-squared:     0.3326310334310908 (original)
'Out of sample' R-squared: 0.29572460427079933 (original)
'In sample' R-squared:     0.33517279824114776 (gen1to5_predict_future)
'Out of sample' R-squared: 0.26262690178799936 (gen1to5_predict_future)


### Question 9:

reading the codes data and the analysis done by chatgpt (mostly) is not super necessary. What I have gleaned from this question is answered in a single sentence at the bottom of this markdown.

- first cell data + analysis:
    - Key Insights:
    Original Model (Overall): The original model trained on the entire dataset has a moderate performance with an in-sample R-squared of 0.378 and an out-of-sample R-squared of 0.351. This shows that the model generalizes reasonably well but might be slightly overfitting to the training data. Generation 1 Model: The model trained on Generation 1 Pokémon performs much better on Generation 1 data (in-sample R-squared of 0.573) but performs very poorly on non-Generation 1 Pokémon (out-of-sample R-squared of 0.112). This suggests that the model is overfitting to the Generation 1 subset and fails to generalize to Pokémon from other generations.

    - Implications:
     The Generation 1 model is not a good predictor for non-Generation 1 Pokémon because it is overfitted to the specific characteristics of Generation 1. It learned patterns that do not transfer well to other Pokémon. The original model (trained on the entire dataset) generalizes slightly better across the dataset but might still benefit from further improvements to handle different subgroups (like different generations of Pokémon). This example highlights the importance of generalization in machine learning: models trained on specific subgroups (e.g., Generation 1 Pokémon) can have poor performance on out-of-sample data if they overfit.

- Second cell data + analysis:

    - Key insights:
    Original Model: In-sample R-squared: 0.378. Out-of-sample R-squared: 0.351. The original model (trained on all generations) explains about 37.8% of the variance in the training data and 35.1% in the test data. There is some overfitting, but overall the model generalizes fairly well across the dataset. Generations 1 to 5 Model: In-sample R-squared: 0.390. Out-of-sample R-squared: 0.234. The model trained on Generations 1 to 5 performs better on the training data (39.0% explained variance) but poorly on Generation 6 data (23.4% explained variance). The substantial drop in performance indicates that this model is overfitting to Generations 1 to 5 and does not generalize well to Generation 6.

    - Implications:
    The Generations 1 to 5 model is not generalizing well to Generation 6, likely due to differences in patterns between the training (Generations 1 to 5) and test (Generation 6) data. This suggests that the model is too specialized to the training data and doesn't capture patterns that apply to Generation 6 Pokémon. The original model (trained on all generations) has better out-of-sample performance compared to the Generations 1 to 5 model, indicating it is less prone to overfitting and generalizes better to unseen data from other generations.
    
    
and so on for cells 3 and 4.

Conclusion: Models in terms of specific generations are not robust in generalising to generations outside their samples. In each case, the original model with data from all generations performs better in terms of generalization and is less prone to overfitting.